File: player/parity/codec.js
/* The parity codec: how a fixture says what JSON cannot (NE-03, plan §6.2).

   A fixture is read by two runtimes — this JS harness and the Swift
   `ForayEngineParity` library — so it has to be plain JSON. Plain JSON cannot
   say NaN, Infinity, -0 or undefined, and the rules the fixtures pin are full of
   them: seam-gap's "a nonsense length collapses to no beat" is a test ABOUT NaN
   and Infinity. So those values travel as tagged objects, and the tags are a
   closed set the schema names (`schema/fixture.schema.json`, `$defs.special`).

   The macros exist for the same two readers. `{"$seg": ["a", 100, 210]}` is
   the queue item every player suite spells `seg("a")`; writing the expansion
   out in every case would make a 300-case family unreadable and would let one
   case's item quietly drift from its neighbours'. The expansions are defined
   HERE and nowhere else on the JS side; the Swift decoder must expand them to
   the same objects, and the `codec` cases in run.test.js pin each one.

   Encoding is only ever applied to what the JS reference RETURNED (an `expect`),
   and decoding only to what a case SUPPLIES (`args`, `setup`, step payloads).
   Macros are therefore legal in inputs only: an `expect` that contained `$seg`
   would be asserting on the macro, not on the rule. */

import fs from "node:fs";
import path from "node:path";

/** The tagged encodings of values JSON cannot carry. Closed. */
export const SPECIAL_NUMBERS = Object.freeze(["NaN", "Infinity", "-Infinity", "-0"]);

/** The input macros. Closed; the schema's `$defs.macro` names the same four. */
export const MACROS = Object.freeze(["$seg", "$ep", "$tts", "$foray"]);

/** A harness error: the CASE is malformed, as opposed to the rule under test
    throwing. The codes are the schema's `$defs.harnessError` enum, so a Swift
    runner reports the same code for the same broken fixture. */
export class HarnessError extends Error {
  constructor(code, message) {
    super(`${code}: ${message}`);
    this.name = "HarnessError";
    this.code = code;
  }
}

const isPlainObject = (v) => v !== null && typeof v === "object" && !Array.isArray(v);

/**
 * Turn a JS value the reference produced into fixture JSON.
 * Throws E_UNENCODABLE for anything a Swift reader could not rebuild — a
 * function, a symbol, a bigint, a class instance with behaviour — rather than
 * letting JSON.stringify drop it and record a fixture that asserts on nothing.
 */
export function encode(value, at = "$") {
  if (value === undefined) return { $undefined: true };
  if (value === null || typeof value === "string" || typeof value === "boolean") return value;
  if (typeof value === "number") {
    if (Number.isNaN(value)) return { $num: "NaN" };
    if (value === Infinity) return { $num: "Infinity" };
    if (value === -Infinity) return { $num: "-Infinity" };
    if (Object.is(value, -0)) return { $num: "-0" };
    return value;
  }
  if (Array.isArray(value)) return value.map((v, i) => encode(v, `${at}[${i}]`));
  if (typeof value === "object") {
    const proto = Object.getPrototypeOf(value);
    if (proto !== Object.prototype && proto !== null) {
      throw new HarnessError("E_UNENCODABLE", `${at} is a ${value.constructor?.name ?? "non-plain object"}; a fixture can only hold plain data`);
    }
    const out = {};
    // Sorted, so a fixture's bytes do not depend on the order a module happened
    // to build its object in — the manifest hashes those bytes.
    for (const k of Object.keys(value).sort()) {
      // An own property holding undefined is dropped, like JSON.stringify does:
      // Swift's decoder cannot tell "absent" from "undefined", so neither may we.
      if (value[k] === undefined) continue;
      out[k] = encode(value[k], `${at}.${k}`);
    }
    return out;
  }
  throw new HarnessError("E_UNENCODABLE", `${at} is a ${typeof value}`);
}

/** The inverse of `encode` for the special tags only. Used on expects before a
    JS-side assertion that wants real values, and on inputs via `expandInputs`. */
export function decodeSpecial(value) {
  if (Array.isArray(value)) return value.map(decodeSpecial);
  if (!isPlainObject(value)) return value;
  const keys = Object.keys(value);
  if (keys.length === 1 && keys[0] === "$num") {
    const tag = value.$num;
    if (!SPECIAL_NUMBERS.includes(tag)) throw new HarnessError("E_BAD_SPECIAL", `unknown $num tag ${JSON.stringify(tag)}`);
    return { NaN: NaN, Infinity: Infinity, "-Infinity": -Infinity, "-0": -0 }[tag];
  }
  if (keys.length === 1 && keys[0] === "$undefined") return undefined;
  const out = {};
  for (const k of keys) out[k] = decodeSpecial(value[k]);
  return out;
}

/* ---------- the macros ---------- */

/** Committed Forays, read once per root. `$foray: "<id>"` is how NE-29j runs
    a rule over every Foray that actually ships, not over a hand-made one. */
const forayCache = new Map();
function committedForay(root, id) {
  if (!forayCache.has(root)) {
    const file = path.join(root, "data", "forays.json");
    const doc = JSON.parse(fs.readFileSync(file, "utf8"));
    forayCache.set(root, new Map((doc.forays ?? []).map((f) => [f.id, f])));
  }
  const foray = forayCache.get(root).get(id);
  if (!foray) throw new HarnessError("E_BAD_MACRO", `$foray names ${JSON.stringify(id)}, which is not in data/forays.json`);
  // A structured clone, so one case mutating its Foray cannot leak into the next.
  return structuredClone(foray);
}

function expandMacro(name, arg, ctx) {
  const tuple = (min, max) => {
    if (!Array.isArray(arg) || arg.length < min || arg.length > max) {
      throw new HarnessError("E_BAD_MACRO", `${name} takes an array of ${min}-${max} elements, got ${JSON.stringify(arg)}`);
    }
    return arg;
  };
  const extra = (v) => {
    if (v === undefined) return {};
    if (!isPlainObject(v)) throw new HarnessError("E_BAD_MACRO", `${name}'s last element must be an object of extra fields`);
    return expandInputs(v, ctx);
  };
  switch (name) {
    // The shape every player suite calls seg(): a real forward slice of a source.
    // Defaults match seam-gap.test.js's `seg(id, 100, 210)`.
    case "$seg": {
      const [id, start = 100, end = 210, more] = tuple(1, 4);
      return { id, kind: "episode", start_sec: decodeSpecial(start), end_sec: decodeSpecial(end), ...extra(more) };
    }
    // An ordinary, unbounded episode.
    case "$ep": {
      const [id, more] = tuple(1, 2);
      return { id, kind: "episode", ...extra(more) };
    }
    // A narration item.
    case "$tts": {
      const [id, more] = tuple(1, 2);
      return { id, kind: "tts", ...extra(more) };
    }
    // A committed Foray by id, or a literal one `{id, title, items}`.
    case "$foray": {
      if (typeof arg === "string") return committedForay(ctx.root, arg);
      if (isPlainObject(arg) && typeof arg.id === "string" && Array.isArray(arg.items)) {
        return { title: "", ...expandInputs(arg, ctx) };
      }
      throw new HarnessError("E_BAD_MACRO", `$foray takes a committed Foray id or {id, title?, items[]}`);
    }
    default:
      throw new HarnessError("E_BAD_MACRO", `unknown macro ${name}`);
  }
}

/**
 * Expand macros and special tags in a case's inputs, producing live JS values.
 * @param {*} value  fixture JSON
 * @param {object} ctx  `{ root }` — the repo root, for `$foray` lookups
 */
export function expandInputs(value, ctx = {}) {
  if (Array.isArray(value)) return value.map((v) => expandInputs(v, ctx));
  if (!isPlainObject(value)) return value;
  const keys = Object.keys(value);
  if (keys.length === 1 && keys[0].startsWith("$")) {
    const k = keys[0];
    if (k === "$num" || k === "$undefined") return decodeSpecial(value);
    if (MACROS.includes(k)) return expandMacro(k, value[k], ctx);
    throw new HarnessError("E_BAD_MACRO", `unknown tag ${k}`);
  }
  const out = {};
  for (const k of keys) {
    if (k.startsWith("$")) throw new HarnessError("E_BAD_MACRO", `a tag (${k}) must be the only key of its object`);
    out[k] = expandInputs(value[k], ctx);
  }
  return out;
}

/** True when `value` (fixture JSON) uses a macro anywhere. Expects may not. */
export function containsMacro(value) {
  if (Array.isArray(value)) return value.some(containsMacro);
  if (!isPlainObject(value)) return false;
  for (const k of Object.keys(value)) {
    if (MACROS.includes(k)) return true;
    if (containsMacro(value[k])) return true;
  }
  return false;
}

File: mobile/plugins/foray-audio/foray-engine-core/Sources/ForayEngineParity/Codec.swift
import Foundation

/// A JavaScript value, as far as a parity case can build one: what a case's
/// `args` MEAN once the runner has expanded them, and what a Swift port's
/// answer is encoded from.
///
/// It exists beside `JSONValue` because the rules the fixtures pin are about
/// values JSON cannot hold. seam-gap's "a nonsense length collapses to no
/// beat" is a test about NaN and Infinity, and "called with nothing at all"
/// is a test about `undefined`. A `FamilyRunner` reads these the way the JS
/// function reads its arguments, then hands its typed Swift port the answer.
public enum JSValue: Equatable {
    case undefined
    case null
    case bool(Bool)
    case number(Double)
    case string(String)
    case array([JSValue])
    case object([String: JSValue])

    /// `value.key` in JavaScript: a missing key, and any member of a
    /// non-object, is `undefined`. (Member access on `null`/`undefined` THROWS
    /// in JavaScript; a runner that can meet one checks `isNullish` first,
    /// because that throw is part of the rule it ports.)
    public subscript(key: String) -> JSValue {
        guard case let .object(fields) = self else { return .undefined }
        return fields[key] ?? .undefined
    }

    /// JavaScript truthiness: `if (value)`.
    public var isTruthy: Bool {
        switch self {
        case .undefined, .null: return false
        case let .bool(value): return value
        case let .number(value): return !(value == 0 || value.isNaN)
        case let .string(value): return !value.isEmpty
        case .array, .object: return true
        }
    }

    public var isNullish: Bool {
        switch self {
        case .undefined, .null: return true
        default: return false
        }
    }

    /// The number, when this is one (`typeof value === "number"`), else nil.
    public var numberValue: Double? {
        if case let .number(value) = self { return value }
        return nil
    }

    public var stringValue: String? {
        if case let .string(value) = self { return value }
        return nil
    }

    /// A plain conversion with NO tag or macro handling: what a JS runner gets
    /// from `JSON.parse` before `expandInputs` touches anything.
    public init(raw: JSONValue) {
        switch raw {
        case .null: self = .null
        case let .bool(value): self = .bool(value)
        case let .number(value): self = .number(value)
        case let .string(value): self = .string(value)
        case let .array(values): self = .array(values.map(JSValue.init(raw:)))
        case let .object(fields): self = .object(fields.mapValues(JSValue.init(raw:)))
        }
    }
}

/// A CASE is malformed, as opposed to the rule under test throwing. The codes
/// are the closed `$defs.harnessError` enum of
/// `player/parity/schema/fixture.schema.json`, the same ones the JS runner
/// raises, so a broken fixture reports the same code in both runtimes.
public struct HarnessError: Error, CustomStringConvertible, Equatable {
    public let code: String
    public let message: String

    public init(_ code: String, _ message: String) {
        self.code = code
        self.message = message
    }

    public var description: String { "\(code): \(message)" }
}

/// The Swift half of `player/parity/codec.js`. Every rule here is that file's,
/// in the same order, and each one says which JS line it mirrors; the
/// `compare` family (which round-trips its inputs through exactly this code on
/// both sides) and the seam-gap family (which leans on `$seg`, `$ep`, `$tts`
/// and every `$num` tag) are what hold the two implementations together.
public enum Codec {
    /// codec.js `SPECIAL_NUMBERS`.
    public static let specialNumbers = ["NaN", "Infinity", "-Infinity", "-0"]
    /// codec.js `MACROS`.
    public static let macros = ["$seg", "$ep", "$tts", "$foray"]

    /// What `$foray: "<id>"` needs: the repo root, to read `data/forays.json`.
    public struct Context {
        public let repoRoot: URL?
        public init(repoRoot: URL?) { self.repoRoot = repoRoot }
    }

    // MARK: encode (codec.js `encode`)

    /// A value the Swift port produced, as fixture JSON. JSON-illegal numbers
    /// become their tags; an object member holding `undefined` is DROPPED, as
    /// `encode` and `JSON.stringify` both drop it (neither reader can tell
    /// "absent" from "undefined", so neither may the comparison).
    public static func encode(_ value: JSValue) -> JSONValue {
        switch value {
        case .undefined: return .object(["$undefined": .bool(true)])
        case .null: return .null
        case let .bool(flag): return .bool(flag)
        case let .string(text): return .string(text)
        case let .number(number): return encode(number)
        case let .array(values): return .array(values.map { encode($0) })
        case let .object(fields):
            var out: [String: JSONValue] = [:]
            for (key, member) in fields where member != .undefined {
                out[key] = encode(member)
            }
            return .object(out)
        }
    }

    public static func encode(_ number: Double) -> JSONValue {
        if number.isNaN { return .object(["$num": .string("NaN")]) }
        if number == .infinity { return .object(["$num": .string("Infinity")]) }
        if number == -.infinity { return .object(["$num": .string("-Infinity")]) }
        if number == 0 && number.sign == .minus { return .object(["$num": .string("-0")]) }
        return .number(number)
    }

    // MARK: decodeSpecial (codec.js `decodeSpecial`)

    /// Tags only, never macros: `{"$num": "NaN"}` -> NaN, `{"$undefined": true}`
    /// -> undefined, recursively. An unknown `$num` tag is E_BAD_SPECIAL.
    public static func decodeSpecial(_ value: JSONValue) throws -> JSValue {
        switch value {
        case let .array(values):
            return .array(try values.map(decodeSpecial))
        case let .object(fields):
            if fields.count == 1, let tag = fields["$num"] {
                switch tag.stringValue {
                case "NaN"?: return .number(.nan)
                case "Infinity"?: return .number(.infinity)
                case "-Infinity"?: return .number(-.infinity)
                case "-0"?: return .number(-0.0)
                default: throw HarnessError("E_BAD_SPECIAL", "unknown $num tag \(tag)")
                }
            }
            // `{"$undefined": <anything>}` is undefined: codec.js checks the key,
            // not the value.
            if fields.count == 1, fields["$undefined"] != nil { return .undefined }
            return .object(try fields.mapValues(decodeSpecial))
        default:
            return JSValue(raw: value)
        }
    }

    // MARK: expandInputs (codec.js `expandInputs`)

    /// A case's inputs as live values: macros expanded, tags decoded.
    public static func expandInputs(_ value: JSONValue, _ ctx: Context) throws -> JSValue {
        switch value {
        case let .array(values):
            return .array(try values.map { try expandInputs($0, ctx) })
        case let .object(fields):
            if fields.count == 1, let key = fields.keys.first, key.hasPrefix("$") {
                if key == "$num" || key == "$undefined" { return try decodeSpecial(value) }
                if macros.contains(key) { return try expandMacro(key, fields[key]!, ctx) }
                throw HarnessError("E_BAD_MACRO", "unknown tag \(key)")
            }
            var out: [String: JSValue] = [:]
            for (key, member) in fields {
                if key.hasPrefix("$") {
                    throw HarnessError("E_BAD_MACRO", "a tag (\(key)) must be the only key of its object")
                }
                out[key] = try expandInputs(member, ctx)
            }
            return .object(out)
        default:
            return JSValue(raw: value)
        }
    }

    static func expandMacro(_ name: String, _ arg: JSONValue, _ ctx: Context) throws -> JSValue {
        func tuple(_ min: Int, _ max: Int) throws -> [JSONValue] {
            guard let items = arg.arrayValue, items.count >= min, items.count <= max else {
                throw HarnessError("E_BAD_MACRO", "\(name) takes an array of \(min)-\(max) elements, got \(arg)")
            }
            return items
        }
        // The trailing "extra fields" object, expanded; absent is {}.
        func extra(_ items: [JSONValue], at index: Int) throws -> [String: JSValue] {
            guard index < items.count else { return [:] }
            guard case .object = items[index] else {
                throw HarnessError("E_BAD_MACRO", "\(name)'s last element must be an object of extra fields")
            }
            guard case let .object(fields) = try expandInputs(items[index], ctx) else { return [:] }
            return fields
        }
        // `{id, kind, ...}` then `...extra`: a later key wins, as a JS spread does.
        func item(_ base: [String: JSValue], _ more: [String: JSValue]) -> JSValue {
            .object(base.merging(more) { _, extraValue in extraValue })
        }
        switch name {
        case "$seg":
            // `const [id, start = 100, end = 210, more] = tuple(1, 4)`: a default
            // applies only to a MISSING element (JSON has no undefined), and the
            // id is taken raw, exactly as the destructuring takes it.
            let items = try tuple(1, 4)
            let start: JSValue = try items.count > 1 ? decodeSpecial(items[1]) : .number(100)
            let end: JSValue = try items.count > 2 ? decodeSpecial(items[2]) : .number(210)
            let more = try extra(items, at: 3)
            return item([
                "id": JSValue(raw: items[0]), "kind": .string("episode"),
                "start_sec": start, "end_sec": end
            ], more)
        case "$ep":
            let items = try tuple(1, 2)
            let more = try extra(items, at: 1)
            return item(["id": JSValue(raw: items[0]), "kind": .string("episode")], more)
        case "$tts":
            let items = try tuple(1, 2)
            let more = try extra(items, at: 1)
            return item(["id": JSValue(raw: items[0]), "kind": .string("tts")], more)
        case "$foray":
            if let id = arg.stringValue { return try committedForay(id, ctx) }
            if case let .object(fields) = arg, fields["id"]?.stringValue != nil, fields["items"]?.arrayValue != nil,
               case let .object(expanded) = try expandInputs(arg, ctx) {
                return item(["title": .string("")], expanded)
            }
            throw HarnessError("E_BAD_MACRO", "$foray takes a committed Foray id or {id, title?, items[]}")
        default:
            throw HarnessError("E_BAD_MACRO", "unknown macro \(name)")
        }
    }

    /// `$foray: "<id>"`: the committed Foray, RAW (codec.js returns a
    /// structured clone of the parsed JSON and expands nothing inside it).
    static func committedForay(_ id: String, _ ctx: Context) throws -> JSValue {
        guard let root = ctx.repoRoot else {
            throw HarnessError("E_BAD_MACRO", "$foray needs the repo root, and this run has none")
        }
        let file = root.appendingPathComponent("data").appendingPathComponent("forays.json")
        let doc = try JSONValue.parse(try Data(contentsOf: file))
        for foray in doc["forays"]?.arrayValue ?? [] where foray["id"]?.stringValue == id {
            return JSValue(raw: foray)
        }
        throw HarnessError("E_BAD_MACRO", "$foray names \"\(id)\", which is not in data/forays.json")
    }
}
